File: shared/src/commonMain/kotlin/org/nihongo/mochi/ui/games/crossword/CrosswordGenerator.kt
package org.nihongo.mochi.ui.games.crossword

import org.nihongo.mochi.domain.words.WordEntry
import kotlin.random.Random

class CrosswordGenerator(
    private val availableWords: List<WordEntry>,
    private val targetWordCount: Int,
    private val mode: CrosswordMode = CrosswordMode.KANAS,
    private val gridSize: Int = 16
) {
    private val grid = Array(gridSize) { Array(gridSize) { "" } }
    private val placedWords = mutableListOf<CrosswordWord>()

    private fun cleanPhonetics(p: String): String {
        return p.split("/")
            .firstOrNull { it.isNotBlank() }
            ?.replace(".", "")
            ?.replace(" ", "")
            ?: ""
    }

    private fun isPureKanji(text: String): Boolean {
        if (text.isEmpty()) return false
        return text.all { char ->
            val code = char.code
            // Kanji range
            (code in 0x4E00..0x9FAF) || (code in 0x3400..0x4DBF)
        }
    }

    fun generate(): Pair<List<CrosswordCell>, List<CrosswordWord>> {
        val candidates = availableWords
            .map { it.copy(phonetics = cleanPhonetics(it.phonetics)) }
            .filter { entry ->
                val solution = if (mode == CrosswordMode.KANJIS) entry.text else entry.phonetics
                val length = solution.length
                val lengthOk = length in 2..8
                
                if (mode == CrosswordMode.KANJIS) {
                    lengthOk && isPureKanji(entry.text)
                } else {
                    lengthOk
                }
            }
            .shuffled()
            .sortedByDescending { if (mode == CrosswordMode.KANJIS) it.text.length else it.phonetics.length }

        if (candidates.isEmpty()) return Pair(emptyList(), emptyList())

        val firstEntry = candidates[0]
        val firstSolution = if (mode == CrosswordMode.KANJIS) firstEntry.text else firstEntry.phonetics
        placeWord(firstEntry, gridSize / 2, (gridSize - firstSolution.length) / 2, true)

        var candidatesIdx = 1
        var attempts = 0
        while (placedWords.size < targetWordCount && candidatesIdx < candidates.size && attempts < 100) {
            val wordEntry = candidates[candidatesIdx]
            if (tryPlaceWord(wordEntry)) {
                attempts = 0
            } else {
                attempts++
            }
            candidatesIdx++
        }

        return finalizeGrid()
    }

    private fun tryPlaceWord(wordEntry: WordEntry): Boolean {
        val word = if (mode == CrosswordMode.KANJIS) wordEntry.text else wordEntry.phonetics
        val possiblePositions = mutableListOf<Triple<Int, Int, Boolean>>()

        for (placed in placedWords) {
            for (i in placed.word.indices) {
                for (j in word.indices) {
                    if (placed.word[i] == word[j]) {
                        if (placed.isHorizontal) {
                            val startRow = placed.row - j
                            val startCol = placed.col + i
                            if (canPlace(word, startRow, startCol, false)) {
                                possiblePositions.add(Triple(startRow, startCol, false))
                            }
                        } else {
                            val startRow = placed.row + i
                            val startCol = placed.col - j
                            if (canPlace(word, startRow, startCol, true)) {
                                possiblePositions.add(Triple(startRow, startCol, true))
                            }
                        }
                    }
                }
            }
        }

        if (possiblePositions.isEmpty()) return false
        val (r, c, horiz) = possiblePositions.random()
        placeWord(wordEntry, r, c, horiz)
        return true
    }

    private fun canPlace(word: String, row: Int, col: Int, isHorizontal: Boolean): Boolean {
        if (row < 0 || col < 0) return false
        if (isHorizontal && col + word.length > gridSize) return false
        if (!isHorizontal && row + word.length > gridSize) return false

        for (i in word.indices) {
            val r = if (isHorizontal) row else row + i
            val c = if (isHorizontal) col + i else col
            val existing = grid[r][c]
            
            if (existing != "" && existing != word[i].toString()) return false
            
            if (existing == "") {
                if (isHorizontal) {
                    if (isOccupied(r - 1, c) || isOccupied(r + 1, c)) return false
                    if (i == 0 && isOccupied(r, c - 1)) return false
                    if (i == word.length - 1 && isOccupied(r, c + 1)) return false
                } else {
                    if (isOccupied(r, c - 1) || isOccupied(r, c + 1)) return false
                    if (i == 0 && isOccupied(r - 1, c)) return false
                    if (i == word.length - 1 && isOccupied(r + 1, c)) return false
                }
            }
        }
        return true
    }

    private fun isOccupied(r: Int, c: Int): Boolean {
        if (r !in 0 until gridSize || c !in 0 until gridSize) return false
        return grid[r][c] != ""
    }

    private fun placeWord(entry: WordEntry, row: Int, col: Int, isHorizontal: Boolean) {
        val solution = if (mode == CrosswordMode.KANJIS) entry.text else entry.phonetics
        
        placedWords.add(
            CrosswordWord(
                number = placedWords.size + 1,
                word = solution,
                kanji = entry.text,
                meaning = "", // On laisse vide, le ViewModel remplira avec la traduction
                phonetics = entry.phonetics,
                row = row,
                col = col,
                isHorizontal = isHorizontal
            )
        )
        for (i in solution.indices) {
            val r = if (isHorizontal) row else row + i
            val c = if (isHorizontal) col + i else col
            grid[r][c] = solution[i].toString()
        }
    }

    private fun finalizeGrid(): Pair<List<CrosswordCell>, List<CrosswordWord>> {
        val cells = mutableListOf<CrosswordCell>()
        val sortedWords = placedWords.sortedWith(compareBy({ it.row }, { it.col }))
        val finalWords = sortedWords.mapIndexed { index, word -> word.copy(number = index + 1) }
        
        for (r in 0 until gridSize) {
            for (c in 0 until gridSize) {
                val char = grid[r][c]
                val wordAtStart = finalWords.find { it.row == r && it.col == c }
                cells.add(
                    CrosswordCell(
                        r = r,
                        c = c,
                        solution = char,
                        isBlack = char == "",
                        number = wordAtStart?.number
                    )
                )
            }
        }
        return Pair(cells, finalWords)
    }
}

File: flutter_app/lib/services/crossword_generator.dart
import 'dart:math';
import '../models/crossword.dart';
import '../repositories/word_repository.dart';

class CrosswordGenerator {
  final List<WordEntry> availableWords;
  final int targetWordCount;
  final CrosswordMode mode;
  final int gridSize;

  CrosswordGenerator({
    required this.availableWords,
    required this.targetWordCount,
    this.mode = CrosswordMode.kanas,
    this.gridSize = 16,
  });

  List<List<String>> _grid = [];
  final List<CrosswordWord> _placedWords = [];
  final Random _random = Random();

  String _cleanPhonetics(String p) {
    return p.split("/").first.replaceAll(".", "").replaceAll(" ", "");
  }

  bool _isPureKanji(String text) {
    if (text.isEmpty) return false;
    return text.runes.all((r) => (r >= 0x4E00 && r <= 0x9FAF) || (r >= 0x3400 && r <= 0x4DBF));
  }

  Map<String, dynamic> generate() {
    _grid = List.generate(gridSize, (_) => List.generate(gridSize, (_) => ""));
    _placedWords.clear();

    final candidates = availableWords
        .map((entry) {
          final cleanP = _cleanPhonetics(entry.phonetics);
          final solution = mode == CrosswordMode.kanjis ? entry.text : cleanP;
          return {'entry': entry, 'solution': solution, 'length': solution.length};
        })
        .where((item) {
          final solution = item['solution'] as String;
          final entry = item['entry'] as WordEntry;
          final bool lengthOk = solution.length >= 2 && solution.length <= 8;
          if (mode == CrosswordMode.kanjis) {
            return lengthOk && _isPureKanji(entry.text);
          }
          return lengthOk;
        })
        .toList();

    if (candidates.isEmpty) return {'cells': <CrosswordCell>[], 'words': <CrosswordWord>[]};

    candidates.shuffle();
    candidates.sort((a, b) => (b['length'] as int).compareTo(a['length'] as int));

    // Placer le premier mot au centre
    final first = candidates[0];
    final firstEntry = first['entry'] as WordEntry;
    _placeWord(firstEntry, gridSize ~/ 2, (gridSize - (first['length'] as int)) ~/ 2, true);

    int candidatesIdx = 1;
    int attempts = 0;
    while (_placedWords.length < targetWordCount && candidatesIdx < candidates.length && attempts < 100) {
      final entry = candidates[candidatesIdx]['entry'] as WordEntry;
      if (_tryPlaceWord(entry)) {
        attempts = 0;
      } else {
        attempts++;
      }
      candidatesIdx++;
    }

    return _finalizeGrid();
  }

  bool _tryPlaceWord(WordEntry wordEntry) {
    final word = mode == CrosswordMode.kanjis ? wordEntry.text : _cleanPhonetics(wordEntry.phonetics);
    final List<Map<String, dynamic>> possiblePositions = [];

    for (var placed in _placedWords) {
      for (int i = 0; i < placed.word.length; i++) {
        for (int j = 0; j < word.length; j++) {
          if (placed.word[i] == word[j]) {
            if (placed.isHorizontal) {
              final startRow = placed.row - j;
              final startCol = placed.col + i;
              if (_canPlace(word, startRow, startCol, false)) {
                possiblePositions.add({'r': startRow, 'c': startCol, 'h': false});
              }
            } else {
              final startRow = placed.row + i;
              final startCol = placed.col - j;
              if (_canPlace(word, startRow, startCol, true)) {
                possiblePositions.add({'r': startRow, 'c': startCol, 'h': true});
              }
            }
          }
        }
      }
    }

    if (possiblePositions.isEmpty) return false;
    final pos = possiblePositions[_random.nextInt(possiblePositions.length)];
    _placeWord(wordEntry, pos['r'], pos['c'], pos['h']);
    return true;
  }

  bool _canPlace(String word, int row, int col, bool isHorizontal) {
    if (row < 0 || col < 0) return false;
    if (isHorizontal && col + word.length > gridSize) return false;
    if (!isHorizontal && row + word.length > gridSize) return false;

    // Check boundary before the start of the word (unconditionally)
    if (isHorizontal) {
      if (_isOccupied(row, col - 1)) return false;
    } else {
      if (_isOccupied(row - 1, col)) return false;
    }

    // Check boundary after the end of the word (unconditionally)
    if (isHorizontal) {
      if (_isOccupied(row, col + word.length)) return false;
    } else {
      if (_isOccupied(row + word.length, col)) return false;
    }

    for (int i = 0; i < word.length; i++) {
      final r = isHorizontal ? row : row + i;
      final c = isHorizontal ? col + i : col;
      final existing = _grid[r][c];

      if (existing != "" && existing != word[i]) return false;

      if (existing == "") {
        // Parallel neighbor checks to keep words isolated
        if (isHorizontal) {
          if (_isOccupied(r - 1, c) || _isOccupied(r + 1, c)) return false;
        } else {
          if (_isOccupied(r, c - 1) || _isOccupied(r, c + 1)) return false;
        }
      }
    }
    return true;
  }

  bool _isOccupied(int r, int c) {
    if (r < 0 || r >= gridSize || c < 0 || c >= gridSize) return false;
    return _grid[r][c] != "";
  }

  void _placeWord(WordEntry entry, int row, int col, bool isHorizontal) {
    final solution = mode == CrosswordMode.kanjis ? entry.text : _cleanPhonetics(entry.phonetics);
    _placedWords.add(CrosswordWord(
      number: _placedWords.length + 1,
      word: solution,
      kanji: entry.text,
      meaning: "", // Rempli par le Provider plus tard
      phonetics: entry.phonetics,
      row: row,
      col: col,
      isHorizontal: isHorizontal,
    ));
    for (int i = 0; i < solution.length; i++) {
      final r = isHorizontal ? row : row + i;
      final c = isHorizontal ? col + i : col;
      _grid[r][c] = solution[i];
    }
  }

  Map<String, dynamic> _finalizeGrid() {
    final List<CrosswordCell> cells = [];
    final sortedWords = List<CrosswordWord>.from(_placedWords)
      ..sort((a, b) => a.row != b.row ? a.row.compareTo(b.row) : a.col.compareTo(b.col));

    final finalWords = sortedWords.asMap().entries.map((e) => e.value.copyWith()).toList();

    for (int r = 0; r < gridSize; r++) {
      for (int c = 0; c < gridSize; c++) {
        final char = _grid[r][c];
        final wordAtStart = finalWords.cast<CrosswordWord?>().firstWhere(
          (w) => w?.row == r && w?.col == c,
          orElse: () => null,
        );
        cells.add(CrosswordCell(
          r: r,
          c: c,
          solution: char,
          isBlack: char == "",
          number: wordAtStart?.number,
        ));
      }
    }
    return {'cells': cells, 'words': finalWords};
  }
}

extension AllExtension<T> on Iterable<T> {
  bool all(bool Function(T) test) {
    for (var element in this) {
      if (!test(element)) return false;
    }
    return true;
  }
}
